rl/q_learner.py

```diff
@@ -0,0 +1,58 @@
+import random
+import numpy as np
+
+class QLearner(object):
+
+  def __init__(self, state_dim,
+                     num_actions,
+                     init_exp=0.5,       # initial exploration prob
+                     final_exp=0.0,      # final exploration prob
+                     anneal_steps=1000, # N steps for annealing exploration 
+                     alpha = 0.2,
+                     discount_factor=0.9, # discount future rewards
+                    ):
+
+    # Q learning parameters
+    self.state_dim       = state_dim
+    self.num_actions     = num_actions
+    self.exploration     = init_exp
+    self.init_exp        = init_exp
+    self.final_exp       = final_exp
+    self.anneal_steps    = anneal_steps
+    self.discount_factor = discount_factor
+    self.alpha           = alpha
+
+    # counters
+    self.train_iteration = 0
+
+    # table of q values
+    self.qtable          = np.random.uniform(low=-1, high=1, size=(state_dim, num_actions))
+
+    
+
+
+  def initializeState(self, state):
+    self.state = state
+    self.action = self.qtable[state].argsort()[-1]
+    return self.action
+
+
+  def nextAction(self, state, reward):
+    if self.exploration > random.random():
+      action = random.randint(0, self.num_actions-1)
+    else:
+      action = self.qtable[state].argsort()[-1]
+
+    self.train_iteration += 1
```

Comment:
Can you move the training code from nextAction() to a new method, like updateModel()? During testing, you want to call nextAction() to only get actions without updating the Q values.